zuul-core/src/main/java/com/netflix/zuul/netty/filter/BaseZuulFilterRunner.java

```diff
@@ -171,18 +180,66 @@ protected final O filter(final ZuulFilter<I, O> filter, final I inMesg) {
             // async filter
-            filter.incrementConcurrency();
-            resumer = new FilterChainResumer(inMesg, filter, snapshot, startTime);
-            filter.applyAsync(inMesg)
-                .observeOn(Schedulers.from(getChannelHandlerContext(inMesg).executor()))
-                .doOnUnsubscribe(resumer::decrementConcurrency)
-                .subscribe(resumer);
+            PerfMark.startTask("FilterRunner.applyAsync", filter.perfmarkTag());
+            try {
+                final Link nettyToSchedulerLink = PerfMark.linkOut();
+                final AtomicReferenceArray<Link> schedulerToNettyLinks = new AtomicReferenceArray<>(3);
+                filter.incrementConcurrency();
+                resumer = new FilterChainResumer(inMesg, filter, snapshot, startTime, schedulerToNettyLinks);
+                filter.applyAsync(inMesg)
+                        .doOnSubscribe(() -> {
+                            PerfMark.startTask("FilterRunner.onSubscribeAsync", filter.perfmarkTag());
+                            try {
+                                PerfMark.linkIn(nettyToSchedulerLink);
+                            } finally {
+                                PerfMark.stopTask("FilterRunner.onSubscribeAsync", filter.perfmarkTag());
+                            }
+                        })
+                        .doOnNext(s -> {
+                            PerfMark.startTask("FilterRunner.onNextAsync", filter.perfmarkTag());
+                            try {
+                                PerfMark.linkIn(nettyToSchedulerLink);
+                                schedulerToNettyLinks.compareAndSet(0, null, PerfMark.linkOut());
+                            } finally {
+                                PerfMark.stopTask("FilterRunner.onNextAsync", filter.perfmarkTag());
+                            }
+                        })
+                        .doOnError(t -> {
+                            PerfMark.startTask("FilterRunner.onErrorAsync", filter.perfmarkTag());
+                            try {
+                                PerfMark.linkIn(nettyToSchedulerLink);
+                                schedulerToNettyLinks.compareAndSet(1, null, PerfMark.linkOut());
+                            } finally {
+                                PerfMark.stopTask("FilterRunner.onErrorAsync", filter.perfmarkTag());
+                            }
+                        })
+                        .doOnCompleted(() -> {
+                            PerfMark.startTask("FilterRunner.onCompletedAsync", filter.perfmarkTag());
+                            try {
+                                PerfMark.linkIn(nettyToSchedulerLink);
+                                schedulerToNettyLinks.compareAndSet(2, null, PerfMark.linkOut());
+                            } finally {
+                                PerfMark.stopTask("FilterRunner.onCompletedAsync", filter.perfmarkTag());
+                            }
+                        })
```

Comment:
Would be nice to move all of these out of this method and into a wrapper. It creates confusion around what the code is doing vs what the benchmarks are doing.

Reply:
Point taken, refactored